src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4487,29 +4429,34 @@ private unsafe char[] GetCanonicalPath(char[] dest, ref int pos, UriFormat forma
                             str = str.Insert(dosPathIdx + _info.Offset.Path - 1, ":");
                         }
 
-                        dest = UriHelper.EscapeString(
+                        UriHelper.EscapeString(
                             str.AsSpan(_info.Offset.Path, _info.Offset.Query - _info.Offset.Path),
-                            dest, ref end,
-                            checkExistingEscaped: !IsImplicitFile, '?', '#');
+                            ref dest, checkExistingEscaped: !IsImplicitFile, '?', '#');
                     }
                     else
                     {
-                        _string.CopyTo(_info.Offset.Path, dest, end, _info.Offset.Query - _info.Offset.Path);
-                        end += (_info.Offset.Query - _info.Offset.Path);
+                        dest.Append(_string.AsSpan(_info.Offset.Path, _info.Offset.Query - _info.Offset.Path));
                     }
                 }
 
                 // On Unix, escape '\\' in path of file uris to '%5C' canonical form.
                 if (!IsWindowsSystem && InFact(Flags.BackslashInPath) && _syntax.NotAny(UriSyntaxFlags.ConvertPathSlashes) && _syntax.InFact(UriSyntaxFlags.FileLikeUri) && !IsImplicitFile)
                 {
-                    dest = UriHelper.EscapeString(new string(dest, pos, end - pos), dest, ref pos, checkExistingEscaped: true, '\\');
-                    end = pos;
+                    // We can't do an in-place escape, create a copy
+                    var copy = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
+                    copy.Append(dest.AsSpan(start, dest.Length - start));
+
+                    dest.Length = start;
+                    ReadOnlySpan<char> copySpan = MemoryMarshal.CreateReadOnlySpan(ref copy.GetPinnableReference(), copy.Length);
```

Comment:
How is this different from `copy.AsSpan()`?

Reply:
Oh, I see your GitHub comment on a similar line below.  It'd be worth a comment in the code.

Reply:
Added comments.
We should be able to Spanify more Uri code now and remove this hack in the future.